### DISCUSSION TITLE
What is the best option for media receive and display

### DISCUSSION THREAD

**Discussion by murphypei** [microsoft/msquic#3313] [Q&A] [answered]
Hi, guys

My usage scenario is that I receive the H.264 data from another machine, I should decode and display it. By reading the [stream Receiving](https://github.com/microsoft/msquic/blob/503e4eef54a3205d4594e16cf299108bc7fe0903/docs/Streams.md#receiving), I think I should receive asynchronously. Does it mean I  should copy the `event->RECEIVE.Buffers[0].Buffer` in the `QUIC_STREAM_EVENT_RECEIVE` event, then decode and display it in another thread and call  [StreamReceiveComplete](https://github.com/microsoft/msquic/blob/503e4eef54a3205d4594e16cf299108bc7fe0903/docs/api/StreamReceiveComplete.md)? Which thread should I call [StreamReceiveSetEnabled](https://github.com/microsoft/msquic/blob/503e4eef54a3205d4594e16cf299108bc7fe0903/docs/api/StreamReceiveSetEnabled.md)? Could I do it synchronously with `QUIC_STATUS_CONTINUE`? Which works better? My encoded data is about 3Mbps.

**Comment by nibanks** [comment, marked as answer]
If you can do it fast enough, then always try to run inline. Otherwise, you probably want to use async receives, which works by returning `QUIC_STATUS_PENDING`, using the buffers/lengths on your own thread, and then calling `StreamReceiveCompete` when you're done with that receive.

**Reply by murphypei** [reply]
thanks.
